src/opm/input/eclipse/Schedule/UDQ/UDQConfig.cpp

```diff
@@ -55,6 +58,73 @@ namespace {
             return s;
         }
     }
+
+    class EvalAssign
+    {
+    public:
+        static EvalAssign field()
+        {
+            return { []() {
+                return [](const auto& assign) { return assign.eval(); };
+            }};
+        }
+
+        static EvalAssign group(const std::size_t    report_step,
+                                const Opm::Schedule& sched)
+        {
+            return { [report_step, &sched]() {
+                return [groups = sched.groupNames(report_step)]
+                    (const auto& assign)
+                {
+                    return assign.eval(groups);
+                };
+            }};
+        }
+
+        static EvalAssign well(const std::size_t    report_step,
+                               const Opm::Schedule& sched)
+        {
+            return { [report_step, &sched]() {
+                return [wells = sched.wellNames(report_step)]
+                    (const auto& assign)
+                {
+                    return assign.eval(wells);
+                };
+            }};
+        }
+
+        static EvalAssign segment(const Opm::UDQContext& context)
+        {
+            return { [&context]() {
+                // Result set should be sized according to total number of
+                // segments.
+                const auto segSet = context.segments();
+
+                return [items = Opm::UDQSet::getSegmentItems(segSet)](const auto& assign)
+                {
+                    return assign.eval(items);
+                };
+            }};
+        }
+
+        Opm::UDQSet operator()(const Opm::UDQAssign& assign) const
+        {
+            if (! this->eval_.has_value()) {
+                this->eval_ = this->create_();
+            }
+
+            return (*this->eval_)(assign);
+        }
```

Comment:
> Here we use the context passed to the factory method. Does that mean the context will change between the creation and the first `operator()` call but not after that?

The context, e.g., the known wells, groups, and segments, will typically not change at all since all of this is run at a single time step, but if it does we have a reference to it instead of a copy.  We use this two-step deferred initialisation process because the `create_()` call could be rather expensive.  It might have to enumerate all wells, or all segments of all multi-segmented wells for that matter, and we prefer not incurring that cost until we know that it's needed.

> Would be cool to see a comment here.

I've tried to explain the reasoning in the updated version of this code.